src/systems/filecoin_markets/storage_market/storage_market_actor_code.go

```diff
@@ -299,6 +301,28 @@ func _rtAbortIfDealInvalidForNewSectorSeal(
 	_rtAbortIfDealExceedsSectorLifetime(rt, dealP, sectorExpiration)
 }
 
+func _rtValidateImmediateCallerDetermineRecipient(rt Runtime, entryAddr addr.Address) addr.Address {
+	if _rtIsStorageMiner(rt, entryAddr) {
+		// Storage miner actor; implied funds recipient is the associated owner address.
+		ownerAddr, workerAddr := _rtGetMinerAccountsAssert(rt, entryAddr)
+		rt.ValidateImmediateCallerInSet([]addr.Address{ownerAddr, workerAddr})
+		return ownerAddr
+	} else {
+		// Account actor (client); funds recipient is just the entry address itself.
+		rt.ValidateImmediateCallerAcceptAnyOfType(actor.BuiltinActorID_Account)
+		return entryAddr
+	}
+}
+
+func _rtIsStorageMiner(rt Runtime, minerAddr addr.Address) bool {
+	codeID, ok := rt.GetActorCodeID(minerAddr)
+	Assert(ok)
+	if !codeID.IsBuiltin() {
+		return false
+	}
+	return (codeID.As_Builtin() == actor.BuiltinActorID_StorageMiner)
```

Comment:
No, this is so brittle! I object in the strongest way. If the storage market needs different behaviour for different kinds of accounts, it should maintain state differentiating those accounts.

Reply:
(Addressed via out-of-band discussions.)